File: bk2ch23p809peopleGroups/ch36p1066peopleGroups/DocumentLister.m


#import "DocumentLister.h"
#import "PeopleLister.h"
#import "AppDelegate.h"

@interface DocumentLister () <UIAlertViewDelegate>
@property (nonatomic, strong) NSMutableArray* files;
@end

@implementation DocumentLister

- (void) viewDidLoad {
    [super viewDidLoad];
    UIBarButtonItem* b =
    [[UIBarButtonItem alloc]
     initWithBarButtonSystemItem:UIBarButtonSystemItemAdd
     target:self action:@selector(doAdd:)];
    self.navigationItem.rightBarButtonItem = b;
    UIBarButtonItem* b2 =
    [[UIBarButtonItem alloc]
     initWithBarButtonSystemItem:UIBarButtonSystemItemRefresh
     target:self action:@selector(doRefresh:)];
    self.navigationItem.leftBarButtonItem = b;
    self.navigationItem.rightBarButtonItem = b2;
    self.title = @"Groups";
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"Cell"];
}

- (void) doRefresh: (id) sender {
    NSFileManager* fm = [NSFileManager new];
    NSURL* docsurl = [fm URLForDirectory:NSDocumentDirectory 
                                inDomain:NSUserDomainMask
                       appropriateForURL:nil create:NO error:nil];
    NSURL* ubiq = [(AppDelegate*)[[UIApplication sharedApplication] delegate] ubiq];
    if (ubiq)
        docsurl = ubiq;
    NSDirectoryEnumerator* dir = [fm enumeratorAtURL:docsurl
                          includingPropertiesForKeys:nil options:0
                                        errorHandler:nil];
    self.files = [NSMutableArray array];
    for (NSURL* aFile in dir) {
        [dir skipDescendants];
        if ([[aFile pathExtension] isEqualToString:@"pplgrp"])
            [self.files addObject:aFile];
    }
    [self.tableView reloadData];
}

- (void) doAdd: (id) sender {
    UIAlertView* av = [[UIAlertView alloc] initWithTitle:@"New Group" message:@"Enter name:" delegate:self cancelButtonTitle:@"Cancel" otherButtonTitles:@"OK", nil];
    av.alertViewStyle = UIAlertViewStylePlainTextInput;
    [[av textFieldAtIndex:0] setAutocapitalizationType:UITextAutocapitalizationTypeWords];
    [av show];
}
    
-(void)alertView:(UIAlertView *)alertView didDismissWithButtonIndex:(NSInteger)buttonIndex {
    if (buttonIndex == [alertView cancelButtonIndex])
        return;
    NSString* name = [alertView textFieldAtIndex:0].text;
    if (!name)
        return;
    if ([name isEqualToString: @""])
        return;
    NSFileManager* fm = [NSFileManager new];
    NSURL* docsurl = [fm URLForDirectory:NSDocumentDirectory 
                                inDomain:NSUserDomainMask appropriateForURL:nil create:NO error:nil];
    NSURL* ubiq = [(AppDelegate*)[[UIApplication sharedApplication] delegate] ubiq];
    if (ubiq)
        docsurl = ubiq;
    name = [name stringByAppendingPathExtension:@"pplgrp"];
    NSURL* url = [docsurl URLByAppendingPathComponent:name];
    if ([fm fileExistsAtPath:url.path])
        return; // TODO: would be nice to alert user to issue here
    PeopleLister* pl = [[PeopleLister alloc] initWithNibName:@"PeopleLister" bundle:nil fileURL:url];
    [self.navigationController pushViewController:pl animated:YES];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self doRefresh:nil];
    
    // good place to create NSMetadataQuery
}


#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    if (!self.files)
        return 0;
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.files count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    
    NSURL* fileURL = (self.files)[indexPath.row];
    cell.textLabel.text = [fileURL.lastPathComponent stringByDeletingPathExtension];
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    
    return cell;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    PeopleLister* pl = [[PeopleLister alloc] initWithNibName:@"PeopleLister" bundle:nil fileURL:(self.files)[indexPath.row]];
    [self.navigationController pushViewController:pl animated:YES];
}

@end

File: bk2ch23p809peopleGroups/ch36p1066peopleGroups/DocumentLister.swift


import UIKit

class DocumentLister: UITableViewController {
    
    var files = [URL]()
    var docsurl : URL {
        var url = NSURL() as URL // stopgap
        let del = UIApplication.shared().delegate
        if let ubiq = (del as! AppDelegate).ubiq {
            url = ubiq
        } else {
            do {
                let fm = FileManager()
                url = try fm.urlForDirectory(.documentDirectory, in: .userDomainMask, appropriateFor: nil, create: false)
            } catch {
                print(error)
            }
        }
        return url
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let b = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(doAdd))
        self.navigationItem.rightBarButtonItems = [b]
        let b2 = UIBarButtonItem(barButtonSystemItem: .refresh, target: self, action: #selector(doRefresh))
        self.navigationItem.leftBarButtonItems = [b2]
        self.title = "Groups"
        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
        
    }
    
    func doRefresh (_:AnyObject?) {
        print("refreshing")
        do {
            let fm = FileManager()
            self.files = try fm.contentsOfDirectory(at:
                self.docsurl, includingPropertiesForKeys: nil)
                .filter {
                    print($0)
                    if fm.isUbiquitousItem(at:$0) {
                        print("trying to download \($0)")
                        try fm.startDownloadingUbiquitousItem(at:$0)
                    }
                    return $0.pathExtension! == "pplgrp"
            }
            self.tableView.reloadData()
        } catch {
            print(error)
        }
    }
    
    func doAdd (_:AnyObject?) {
        let av = UIAlertController(title: "New Group", message: "Enter name:", preferredStyle: .alert)
        av.addTextField {$0.autocapitalizationType = .words}
        av.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        av.addAction(UIAlertAction(title: "OK", style: .default) {
            _ in
            guard let name = av.textFields![0].text where !name.isEmpty else {return}
            let url = try! self.docsurl.appendingPathComponent((name as NSString).appendingPathExtension("pplgrp")!)
            // really should check to see if file by this name exists
            let pl = PeopleLister(fileURL: url)
            self.navigationController!.pushViewController(pl, animated: true)
        })
        self.present(av, animated: true)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.doRefresh(nil)
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.files.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier:"Cell", for: indexPath)
        let fileURL = self.files[indexPath.row]
        cell.textLabel!.text = (fileURL.lastPathComponent! as NSString).deletingPathExtension
        cell.accessoryType = .disclosureIndicator
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let pl = PeopleLister(fileURL: self.files[indexPath.row])
        self.navigationController!.pushViewController(pl, animated: true)
    }

    
}
